### ISSUE TITLE
Webpack 5 memory leak in watch mode

### ISSUE THREAD

**Issue by Geksanit** [webpack/webpack#12771]
# Bug report

**What is the current behavior?**
When I make a change in my web app and Webpack compiles, memory usage seems to increase. If I keep making changes and Webpack keeps re-compiling, memory usage will continue to climb. In a large project I work in, it takes about 10 compilations until Webpack crashes with a "Allocation failed - JavaScript heap out of memory" exception.

**If the current behavior is a bug, please provide the steps to reproduce.**
Please see the minimal repo I created to reproduce the issue: https://github.com/Geksanit/wp5-memory-leak.
1 yarn start
2 save any ts/tsx files without changes
![dev-tools-snapshots](https://user-images.githubusercontent.com/31949379/108956295-1429df00-76a2-11eb-8f83-e960e02c5db0.JPG)

**What is the expected behavior?**
I would expect some kind of garbage collection to happen in between compilations.

**Other relevant information:**
webpack version:  5.24.1
Node.js version: 12.19.0
Operating System: Win10
Additional tools: yarn 1.22.10

**Comment by piecyk** [comment]
I'm also seeing this, running webpack `5.24.2`

**Comment by piecyk** [comment]
Trying to pin this but nothing concrete, till then kinda suspicious looks Compilation 🤔 

<img width="1304" alt="Screenshot 2021-02-26 at 11 55 32" src="https://user-images.githubusercontent.com/82964/109291847-c809a680-7829-11eb-9231-cf4e267fb239.png">

Compering snapshot 2 vs 3 

<img width="1286" alt="Screenshot 2021-02-26 at 11 56 02" src="https://user-images.githubusercontent.com/82964/109291899-dfe12a80-7829-11eb-864a-74bbed41d825.png">

I think it's connected with new asset/resource and mini-css-extract-plugin

running webpack v5.24.2, node v12.14.1

**Comment by alexander-akait** [comment]
@piecyk in original issue, repo don't use `mini-css-extract-plugin`, try to disable it for debug

**Comment by alexander-akait** [comment]
@piecyk What is value `devtool`?

**Comment by piecyk** [comment]
>  in original issue, repo doesn't use mini-css-extract-plugin, try to disable it for debug

yeah, trying to create minimal repo with minimal configuration that we can work on. 

> What is value devtool

devtool: false

**Comment by alexander-akait** [comment]
@piecyk Can you use `Comparsion` with the first/second snapshots (and sort using `Size Delta` column)?

**Comment by piecyk** [comment]
Hmm something like, also will try to look on this more over the weekend.

First example without mini-css-extract-plugin, rather common setup, babel, react, postcss, html-webpack-plugin, 
second with latest  v1.3.9, actions are identical and snapshots are taken after each step compilation finished 

1. after cold start
2. change file near the top 
3. revert changed file 

<img width="1295" alt="Screenshot 2021-02-26 at 15 17 26" src="https://user-images.githubusercontent.com/82964/109312265-1f6a3f80-7847-11eb-8218-1c7af0aa2640.png">

Then with latest `mini-css-extract-plugin`, surpassing is the impact on memory 

<img width="1280" alt="Screenshot 2021-02-26 at 15 26 25" src="https://user-images.githubusercontent.com/82964/109313069-15950c00-7848-11eb-814f-ed7f4c9113df.png">

**Comment by alexander-akait** [comment]
I think here bug not with `mini-css-extract-plugin`, here bug between memory cache (can you try `cache: false`?) and assets and child compilations, hm, but need snapshots/reproducible repo to make sure of this

**Comment by sokra** [comment]
I can't reproduce that. It doesn't leak memory for me...

But I can help you to use the dev tools and find it yourself:

1. Run a few rebuilds to warmup everything
2. Take 3 memory snapshots for 3 rebuilds. Click "Collect garbage" before taking the snapshots.
3. Select snapshot 3.
4. Choose "Objects allocated between Snapshot 1 and Snapshot 2". That will only show object the are leaking from the first rebuild.
5. There should be only system objects now. It's never empty, but there shouldn't be any webpack object. Look for Compilation or Module. That must not be there.
6. Check Retainers to find the cause. Only follow chains where "Distance" is decreasing. The others will not lead to the cause.

![image](https://user-images.githubusercontent.com/1365881/109325082-ce157c80-7855-11eb-8cac-4888a836509f.png)

**Comment by Geksanit** [comment]
I can't reproduce  that with rebuilding one file. But I found this.

1. Run a few rebuilds (App.tsx and index.ts)
2. Click "Collect garbage" and "Take Snapshot"
3. Save App.tsx and index.ts
4. Repeat.
![image](https://user-images.githubusercontent.com/31949379/109393112-9b9d7980-7952-11eb-8040-e05ce1cb2e17.png)

**Comment by alexander-akait** [comment]
@Geksanit can you try to change `logo.svg` (just adding newline and remove it) and record it after ~10/25/50 changes changes

**Comment by Geksanit** [comment]
> @Geksanit can you try to change `logo.svg` (just adding newline and remove it) and record it after ~10/25/50 changes

![image](https://user-images.githubusercontent.com/31949379/109410491-ad703280-79cd-11eb-946a-6cef0b9f9504.png)

**Comment by alexander-akait** [comment]
hm, no leak...

**Comment by kanoshin** [comment]
@sokra @alexander-akait there is no memory leak if you keep editing the same file but it is reproducible if you try to edit different files between heap snapshots. 

Try this:
1. Run build and modify `index.tsx` file
2. Create first snapshot
3. Modify `App.tsx`
4. Create second snapshot
5. Modify `App.css`
6. Create third snapshot
7. Now there will be webpack related objects in the snapshot diff

<img width="1201" alt="Screen Shot 2021-03-09 at 11 39 17 PM" src="https://user-images.githubusercontent.com/11817360/110593648-b204bb80-8130-11eb-8ec9-85e40691d9f7.png">

**Comment by sokra** [comment]
Yes, we should not store the Compilation in AssetGenerator. Let me fix that and recheck...

**Comment by kanoshin** [comment]
@sokra there is still a leak if I don't use AssetGenerator at all:

<img width="1920" alt="Screen Shot 2021-03-11 at 11 42 41 AM" src="https://user-images.githubusercontent.com/11817360/110845065-06a65480-825f-11eb-9c51-4da2ba6286c0.png">

Do you want me to create a new issue for that? Here is what I did to reproduce without AssetGenerator: https://github.com/kanoshin/wp5-memory-leak/commit/67fb5167067c2ef6135d06916ea5371343997aa4

**Comment by sokra** [comment]
Hmm yeah I see. That's the weird backward-compat stuff...

**Comment by sokra** [comment]
Ok I looked into that a bit deeper and there is a pretty serious memory leak that happens when removing modules from the compilation. e. g. removing an `import`.
But it's not that easy to fix.

One problem are backward-compat back references to objects one should not have access too. That's easy to fix by removing the backward-compat references when the next Compilation starts. In the next major they will be removed anyway.

Another problem is that modules in cache keep referencing the Compilation they were created in (indirectly e. g. through the parser), which then references a lot other objects. We need to move such references (like parser, generator, etc) from the Module to the ModuleGraph and give the Module access to the ModuleGraph at places where it want to access it's references. And to make it backward-compatible we need to add a lot getters for the old references.

**Comment by sokra** [comment]
Interesting thing I figured out:

Using the devtools sometimes show fake memory leaks. Seems like the something in devtools is holding some references which prevents that some WeakMap references to be collected. I think it's either the Source panel or the memory dumps themselves. Creating heapdumps via API fixes that problem...

Anyway I found a memory leak that happens when modules are no longer referenced by the compilation, but stay in cache. Here they keep some references to the old compilation and keep it alive. That should not happen and will be fixed by #12894.
